### PR TITLE
Adjust the `gradcheck` tests for `ComplexTensor`

### PR DESCRIPTION
**Pull request by hameerabbasi**
The existing `gradcheck` tests appear to have some bugs when it comes to dealing with a `TorchDispatchMode`. In this PR, I've hand-rolled some tests that manually use `torch.autograd.gradcheck.gradcheck`; and all of these work as expected.

### CODE REVIEW

**Comment by ezyang** [comment]
Endorsing, but @albanD you might wanna take a look

**Review by amjames — APPROVED** [review]

File: torch/_subclasses/complex_tensor/_ops/aten.py [hunk 1]
```diff
@@ -699,7 +699,11 @@ def linalg_vector_norm_impl(self: ComplexTensor, *args, **kwargs) -> torch.Tenso
 
 @register_force_test(aten.copy_)
 def copy__impl(self: ComplexTensor, src, *args, **kwargs):
-    self_re, self_im = split_complex_tensor(self)
+    if not self.dtype.is_complex:
```
Comment (albanD) [inline torch/_subclasses/complex_tensor/_ops/aten.py:708]: Either this condition will never trigger or the typing above is wrong?

File: test/complex_tensor/utils.py [hunk 2]
```diff
@@ -149,20 +165,64 @@ def check_consistency(
...
+                        return op(
+                            subclass_sample.input,
+                            *subclass_sample.args,
+                            **subclass_sample.kwargs,
+                        )
+
+                    self.assertSameResult(expected, actual, **kwargs)
+                case Variant.GradCheck:
+                    input_list = []
+                    # Mark all applicable tensors with `requires_grad=True`
```
Comment (albanD) [inline test/complex_tensor/utils.py]: You can pass requires_grad=True to op.sample_inputs()

File: test/complex_tensor/utils.py [hunk 3]
```diff
@@ -149,20 +165,64 @@ def check_consistency(
...
+                            if not isinstance(x, torch.Tensor):
+                                return x
+                            return next(it)
+
+                        input_sample = sample_input.transform(f)
+                        return op(
+                            input_sample.input,
+                            *input_sample.args,
+                            **input_sample.kwargs,
+                        )
```
Comment (albanD) [inline test/complex_tensor/utils.py]: Why all of this? Inputs that are not Tensors that require grad are just ignored by gradcheck

File: test/complex_tensor/utils.py [hunk 4]
```diff
@@ -149,20 +165,64 @@ def check_consistency(
...
+                    subclass_sample = sample_input.transform(transform_fn)
+
+                    def actual(subclass_sample=subclass_sample):
+                        return op(
+                            subclass_sample.input,
+                            *subclass_sample.args,
+                            **subclass_sample.kwargs,
+                        )
+
+                    self.assertSameResult(expected, actual, **kwargs)
```
Comment (albanD) [inline test/complex_tensor/utils.py]: nit: this was here before so no need to change.
But what's the point of wrapping all of this in functions to check the result rather than just calling op inline and assertEqual ?

Comment (hameerabbasi) [reply test/complex_tensor/utils.py]: I introduced a match statement inside which this is nested now; so there's an extra layer of indentation in the diff. Previously, this fn wasn't being called at all for `Variant.GradCheck`, now it is, and it's different from the other two tests.

File: test/complex_tensor/utils.py (hunk 2 above)
Comment (hameerabbasi) [reply test/complex_tensor/utils.py]: I tried this (in combination with simplifying according to https://github.com/pytorch/pytorch/pull/168248#discussion_r2556569173)

<details><summary>Diff applied</summary>

```diff
diff --git a/test/complex_tensor/utils.py b/test/complex_tensor/utils.py
index 03da1fe8b27..478f1fd421b 100644
--- a/test/complex_tensor/utils.py
+++ b/test/complex_tensor/utils.py
@@ -161,7 +161,7 @@ class TestCase(PytorchTestCase):
             if extra_info.matches(test_info):
                 kwargs = extra_kw
                 break
-        sample_inputs = op.sample_inputs(device, dtype)
+        sample_inputs = op.sample_inputs(device, dtype, requires_grad=(variant == Variant.GradCheck))
         transform_fn = TRANSFORM_FUNCS[variant]
 
         for sample_input in sample_inputs:
@@ -186,35 +186,35 @@ class TestCase(PytorchTestCase):
 
                     self.assertSameResult(expected, actual, **kwargs)
                 case Variant.GradCheck:
-                    input_list = []
-                    # Mark all applicable tensors with `requires_grad=True`
-                    # Come up with a list of applicable tensors.
-                    subclass_sample = sample_input.transform(
-                        functools.partial(transform_fn, input_list=input_list)
-                    )
-
-                    # The function to gradcheck is the op, replacing the appropriate tensors
-                    # with the input ones
-                    def func(*t: torch.Tensor):
-                        it = iter(t)
-
-                        def f(x):
-                            if not isinstance(x, torch.Tensor):
-                                return x
-                            return next(it)
-
-                        input_sample = sample_input.transform(f)
-                        return op(
-                            input_sample.input,
-                            *input_sample.args,
-                            **input_sample.kwargs,
-                        )
+                    # input_list = []
+                    # # Mark all applicable tensors with `requires_grad=True`
+                    # # Come up with a list of applicable tensors.
+                    # subclass_sample = sample_input.transform(
+                    #     functools.partial(transform_fn, input_list=input_list)
+                    # )
+
+                    # # The function to gradcheck is the op, replacing the appropriate tensors
+                    # # with the input ones
+                    # def func(*t: torch.Tensor):
+                    #     it = iter(t)
+
+                    #     def f(x):
+                    #         if not isinstance(x, torch.Tensor):
+                    #             return x
+                    #         return next(it)
+
+                    #     input_sample = sample_input.transform(f)
+                    #     return op(
+                    #         input_sample.input,
+                    #         *input_sample.args,
+                    #         **input_sample.kwargs,
+                    #     )
 
                     # Do the actual gradcheck
                     self.assertTrue(
                         gradcheck(
-                            func,
-                            input_list,
+                            functools.partial(op, **sample_input.kwargs),
+                            (sample_input.input, *sample_input.args),
                             raise_exception=True,
                             fast_mode=True,
                             check_batched_grad=True,
```
</details>

<details><summary>Failures</summary>

```
================================================= short test summary info =================================================
FAILED [0.0342s] test/complex_tensor/test_complex_tensor.py::TestComplexBwdGradientsCPU::test_fn_grad_cat_cpu_complex128 - Exception: gradcheck expects at least one input tensor to require gradient, but none of the them have requires_grad=True.
FAILED [0.0346s] test/complex_tensor/test_complex_tensor.py::TestComplexBwdGradientsCPU::test_fn_grad_stack_cpu_complex128 - Exception: gradcheck expects at least one input tensor to require gradient, but none of the them have requires_grad=True.
FAILED [0.0344s] test/complex_tensor/test_complex_tensor.py::TestComplexBwdGradientsCUDA::test_fn_grad_cat_cuda_complex128 - Exception: gradcheck expects at least one input tensor to require gradient, but none of the them have requires_grad=True.
FAILED [0.0344s] test/complex_tensor/test_complex_tensor.py::TestComplexBwdGradientsCUDA::test_fn_grad_stack_cuda_complex128 - Exception: gradcheck expects at least one input tensor to require gradient, but none of the them have requires_grad=True.
```
</details>

It seems like `sample_ops` doesn't do `requires_grad=True` for `stack` and `cat`. My hand-rolled solution works in these cases.

File: torch/_subclasses/complex_tensor/_ops/aten.py (hunk 1 above)
Comment (hameerabbasi) [reply torch/_subclasses/complex_tensor/_ops/aten.py:708]: Nice catch, fixed in https://github.com/pytorch/pytorch/pull/168248/commits/4333f86c1d2ca4a6e227ef666c9222c89de3ed4c.

File: test/complex_tensor/utils.py (hunk 3 above)
Comment (hameerabbasi) [reply test/complex_tensor/utils.py]: See https://github.com/pytorch/pytorch/pull/168248#discussion_r2558789342

File: test/complex_tensor/utils.py (hunk 2 above)
Comment (albanD) [reply test/complex_tensor/utils.py]: So the filtering does work fine. And we just need to make sure cat/stack test do generate inputs that require gradients. Updating their sample function should do the trick.
That would be preferable to fix this for everyone rather than this etra logic.

File: test/complex_tensor/utils.py (hunk 4 above)
Comment (albanD) [reply test/complex_tensor/utils.py]: Not sure what you mean? I agree this was there before.
My question is only about why this design?

Comment (hameerabbasi) [reply test/complex_tensor/utils.py]: You're right, this could be in another function, I've done that in https://github.com/pytorch/pytorch/pull/168248/commits/028508aef59d1452ed5c184645e2d0c034947af9.

File: test/complex_tensor/utils.py (hunk 2 above)
Comment (hameerabbasi) [reply test/complex_tensor/utils.py]: Done in https://github.com/pytorch/pytorch/pull/168248/commits/028508aef59d1452ed5c184645e2d0c034947af9.

**Comment by hameerabbasi** [comment]
@albanD A re-review would be appreciated, turns out that `SampleInput.transform` runs the transformation in a `no_grad` context which makes it lose the `requires_grad` flag.

File: test/complex_tensor/utils.py [hunk 5]
```diff
@@ -123,35 +138,54 @@ def assertSameResult(
...
 
-    def check_consistency(
-        self, device: str, dtype, op: OpInfo, variant: Variant
-    ) -> None:
+    def conditional_skip(self, desc: Descriptor) -> None:
+        try:
+            from .test_complex_tensor import SKIPS
+        except ImportError:
+            from test_complex_tensor import SKIPS
+        for xfail_info, reason in SKIPS.items():
```
Comment (albanD) [inline test/complex_tensor/utils.py:132]: If these are xfail, we should xfail, not skip

File: test/complex_tensor/utils.py [hunk 6]
```diff
@@ -49,9 +50,23 @@ def _as_complex_dtensor(arg: torch.Tensor | Any) -> torch.Tensor | Any:
...
 
 
+def _as_gradcheck_tensor(
+    arg: torch.Tensor | Any, input_list: list[torch.Tensor]
+) -> torch.Tensor | Any:
+    if not isinstance(arg, torch.Tensor):
+        return arg
+    arg = _as_complex_tensor(arg)
+    if arg.dtype.is_floating_point or arg.dtype.is_complex:
+        arg = arg.requires_grad_()
```
Comment (albanD) [inline test/complex_tensor/utils.py]: Why do you need this? Aren't you not using transform explicitly to avoid having to do this?

File: torch/_subclasses/complex_tensor/_ops/aten.py [hunk 7]
```diff
@@ -698,8 +698,17 @@ def linalg_vector_norm_impl(self: ComplexTensor, *args, **kwargs) -> torch.Tenso
...
-def copy__impl(self: ComplexTensor, src, *args, **kwargs):
-    self_re, self_im = split_complex_tensor(self)
+def copy__impl(
+    self: ComplexTensor | torch.Tensor,
+    src: ComplexTensor | torch.Tensor,
+    *args,
+    **kwargs,
+) -> ComplexTensor | torch.Tensor:
+    if not self.dtype.is_complex:
+        src_re, src_im = split_complex_arg(src)
```
Comment (albanD) [inline torch/_subclasses/complex_tensor/_ops/aten.py:712]: nit: you can keep the warning that exists in the current complex impl here

**Review by albanD — APPROVED** [review]
Sounds ok.

File: test/complex_tensor/utils.py (hunk 5 above)
Comment (hameerabbasi) [reply test/complex_tensor/utils.py:132]: This code was here before, just moved to a more convenient place. Turns out that using `xfails` requires some engineering effort; you can't do it inside the test, it has to be at collection time.

File: test/complex_tensor/utils.py (hunk 6 above)
Comment (hameerabbasi) [reply test/complex_tensor/utils.py]: Yes we have replaced this with `OpInfo.sample(..., requires_grad=True)`, it's dead code now. Removed in https://github.com/pytorch/pytorch/pull/168248/commits/7ca667fdd51d89a754f28c3177803dab129f2bea.

File: torch/_subclasses/complex_tensor/_ops/aten.py (hunk 7 above)
Comment (hameerabbasi) [reply torch/_subclasses/complex_tensor/_ops/aten.py:712]: Done in https://github.com/pytorch/pytorch/pull/168248/commits/bc1b376b3b86a4468093dbeaf99d8484d42ffd9f.
